### PR TITLE
BETA Stackstorm Enterprise HA cluster with K8s & Helm

### PR DESCRIPTION
**Pull request by armab**
Closes StackStorm/k8s-st2#61

Preparing StackStorm Enterprise HA cluster documentation with Docker, K8s and Helm for upcoming `v2.9`.

----

Still missing some references, as well as stabilized "Custom st2 packs" story (StackStorm/k8s-st2/pull/78) after which we can extract Helm chart from https://github.com/StackStorm/k8s-st2 into a https://github.com/stackstorm/stackstorm-enterprise-ha and make it public.

### CODE REVIEW

File: docs/source/install/ewc_ha.rst [hunk 1]
```diff
@@ -0,0 +1,279 @@
+|bwc| HA Cluster - BETA
+=======================
+
+|bwc| (EWC) is the commercial version of the StackStorm automation platform. EWC adds priority
+support, advanced features such as fine-tuned access control, LDAP, and Workflow Designer. To
+learn more about |bwc|, get an evaluation license, or request a quote, visit `extremenetworks.com/product/workflow-composer
+<https://www.extremenetworks.com/product/workflow-composer/>`_.
+
+This document provides an installation blueprint for a Highly Availabile StackStorm Enterprise (|bwc|) cluster
+based on `Kubernetes <https://kubernetes.io/>`__, a container orchestration platform with a planet scale.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: "with a planet scale"? If you choose to use this term, perhaps replace with "at planet scale" instead? @LindsayHill thoughts?

File: docs/source/install/ewc_ha.rst [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+The cluster deploys minimum 2 replicas for each component of StackStorm microservices for redundancy and reliability,
+as well as configures backends like MongoDB HA Replicaset, RabbitMQ HA and etcd cluster that st2 relies on for database,
+communication bus, and distributed coordination respectively. That raises a fleet of more than a ``30`` pods in summ.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: Not sure what you mean by "in summ"? Either drop it, or clarify.

File: docs/source/install/ewc_ha.rst [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+The source code for K8s resource templates is available as a GitHub repo (TODO):
+`StackStorm/stackstorm-enterprise-ha <https://github.com/StackStorm/stackstorm-enterprise-ha>`_.
+
+.. note::
+
+    **Warning! Beta quality!**
+
+    As this deployment method available in beta version, documentation and code may be substantially modified and refactored.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...method is available...

File: docs/source/install/ewc_ha.rst [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+    Request a 90-day free trial at https://stackstorm.com/#product
+
+Once the deployment is finished, it'll show you first steps how to start working with the new cluster via WebUI or st2 client:
+
+.. figure :: /_static/images/helm-chart-notes.png
+    :align: center
+
+
+The installation uses some unsafe defaults which are recommended to change thoughtfully for production use via Helm ``values.yaml``.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...which we recommend you change...

File: docs/source/install/ewc_ha.rst [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+- number of replicas for each component
+- st2 auth secrets
+- st2.conf settings
+- RBAC roles, assignments and mappings
+- custom st2 pack configs
+- st2web SSL certificate
+- SSH private key
+- K8s resources and settings to control pod/deployment placement
+- configuration for Mongo, RabbitMQ clusters
+
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: Consider adding note about configuring st2 packs (pack configs as well as image repo config)

File: docs/source/install/ewc_ha.rst [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+  kubectl logs -l release=<release-name>,tier=backend
+
+
+Custom st2 packs
+----------------
+To follow the stateless model, shipping custom st2 packs is now part of the deployment process.
+It means that ``st2 pack install`` won't work in a distributed environment and you have to bundle all the
+required packs into a Docker image that you could codify, version and distribute via preferred Docker registry,
+is it private or not. The responsibility of such Docker image is to hold pack content and their virtualenvs.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...registry, whether private or not.

File: docs/source/install/ewc_ha.rst [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+So custom st2 pack Docker image you have to build is just read-only dirs that are shared with the corresponding
+st2 services in a cluster.
+
+For the convenience we created new ``st2-pack-install <pack1> <pack2> <pack3>`` command
+that'll help to install custom packs during the Docker build process without relying on DB and MQ connection.
+
+Helm chart brings helpers to simplify this experience like `stackstorm/st2pack-builder <https://hub.docker.com/r/stackstorm/st2packs-builder/>`_
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: https://hub.docker.com/r/stackstorm/st2packs:builder and https://hub.docker.com/r/stackstorm/st2packs:runtime

However, the user will build st2packs image using the simple `Dockerfile` in `stackstorm-enterprise-ha` git repo.

File: docs/source/install/ewc_ha.rst [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+For more detailed instructions see `StackStorm/stackstorm-enterprise-ha#Installing packs in the cluster <https://github.com/StackStorm/stackstorm-enterprise-ha#Installing-packs-in-the-cluster>`_.
+
+.. note::
+  There is an alternative approach, - sharing pack content via read-write-many NFS (Network File System) as :doc:`/reference/ha` recommends.
+  As beta is in progress and both methods have their pros and cons, we'd like to hear your feedback and which way would work better for you.
+
+Components
+----------
+For HA reasons, by default and at a minimum StackStorm K8s cluster deploys more than a ``30`` pods in total.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...more than ``30``...

File: docs/source/install/ewc_ha.rst [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+Relies on ``etcd`` for coordination. This is likely the first thing to lift if you have a lot of actions
+to execute per time period in your StackStorm cluster.
+
+st2garbagecollector
+___________________
+Service that cleans up old executions and other operations data based on setup configurations.
+Having ``1`` st2garbagecollector replica for K8s Deployment is enough, considering its periodic execution nature.
+By default this process does nothing and needs to be configured in st2.conf settings (via ``values.yaml``).
+Purging stale data can significantly improve cluster abilities to perform faster and so it's recommended to configure st2garbagecollector in production.
+
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: Worth mentioning docker-registry and kube-registry-proxy here? They're optional services... based entirely on st2.packs config.

File: docs/source/install/index.rst [hunk 10]
```diff
@@ -29,6 +29,9 @@ Here's an overview of the options:
 * **Puppet Module:** For Puppet users, check this :doc:`/install/puppet` for
   installing |st2|. A robust and idempotent method of installing and configuring |st2|.
 * **Docker:** |st2| is now supported on Docker - check out our :doc:`docker` instructions.
+* **High Availability** Entrusting business critical automation tasks to a system like StackStorm
+  leads to a higher demands from that system. StackStorm can run in a HA mode to ensure these needs.
```
Comment (warrenvw) [inline docs/source/install/index.rst]: Consider "...leads to higher demands on that system."

File: docs/source/install/ewc_ha.rst (hunk 9 above)
Comment (armab) [reply docs/source/install/ewc_ha.rst]: Per PR comment I'm still waiting for https://github.com/StackStorm/k8s-st2/pull/78 to be finished.

Just took some descriptions from that PR and added new section in https://github.com/StackStorm/st2docs/pull/795/commits/806eebb0af92519807b121e29c8aa134bfd792a9 for private docker registry.

**Comment by armab** [comment]
@warrenvw Thanks for the comments :+1: 
I've addressed your review points, but probably would be good to take another look.

File: docs/source/install/ewc_ha.rst [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+|bwc| (EWC) is the commercial version of the StackStorm automation platform. EWC adds priority
+support, advanced features such as fine-tuned access control, LDAP, and Workflow Designer. To
+learn more about |bwc|, get an evaluation license, or request a quote, visit `extremenetworks.com/product/workflow-composer
+<https://www.extremenetworks.com/product/workflow-composer/>`_.
+
+This document provides an installation blueprint for a Highly Availabile StackStorm Enterprise (|bwc|) cluster
+based on `Kubernetes <https://kubernetes.io/>`__, a container orchestration platform at planet scale.
+
+The cluster deploys minimum 2 replicas for each component of StackStorm microservices for redundancy and reliability,
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...deploys a minimum of 2 replicas...

File: docs/source/install/ewc_ha.rst [hunk 12]
```diff
@@ -0,0 +1,286 @@
+|bwc| HA Cluster - BETA
+=======================
+
+|bwc| (EWC) is the commercial version of the StackStorm automation platform. EWC adds priority
+support, advanced features such as fine-tuned access control, LDAP, and Workflow Designer. To
+learn more about |bwc|, get an evaluation license, or request a quote, visit `extremenetworks.com/product/workflow-composer
+<https://www.extremenetworks.com/product/workflow-composer/>`_.
+
+This document provides an installation blueprint for a Highly Availabile StackStorm Enterprise (|bwc|) cluster
+based on `Kubernetes <https://kubernetes.io/>`__, a container orchestration platform at planet scale.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: Sorry.. still doesn't read quite right. Perhaps "for planet scale"?

File: docs/source/install/ewc_ha.rst [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+
+The cluster deploys minimum 2 replicas for each component of StackStorm microservices for redundancy and reliability,
+as well as configures backends like MongoDB HA Replicaset, RabbitMQ HA and etcd cluster that st2 relies on for database,
+communication bus, and distributed coordination respectively. That raises a fleet of more than a ``30`` pods total.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...more than ``30`` pods total.

File: docs/source/install/ewc_ha.rst [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+* `Helm <https://docs.helm.sh/using_helm/#install-helm>`__, the K8s package manager and `Tiller <https://docs.helm.sh/using_helm/#initialize-helm-and-install-tiller>`_
+* Enough computing resources for production use, respecting :doc:`/install/system_requirements`
+
+Usage
+-----
+This document assumes some basic knowledge of Kubernetes and Helm.
+Please refer to `K8s <https://kubernetes.io/docs/home/>`__ and `Helm <https://docs.helm.sh/>`__
+documentation if you find any difficulties using these tools.
+
+However here are some minimal instructions to get started.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: However,

File: docs/source/install/ewc_ha.rst [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+
+.. figure :: /_static/images/helm-chart-notes.png
+    :align: center
+
+
+The installation uses some unsafe defaults which we recommend you change thoughtfully for production use via Helm ``values.yaml``.
+
+Helm values.yaml
+________________
+Helm package ``stackstorm-enterprise-ha`` comes with the default settings in ``values.yaml``.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...comes with default settings...

File: docs/source/install/ewc_ha.rst [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+Upgrading
+_________
+Once you make any changes to Helm values, upgrade the cluster:
+
+.. code-block:: bash
+
+  helm repo update
+  helm upgrade <release-name> stackstorm/stackstorm-enterprise-ha
+
+It will redeploy components which were affected by the change, taking care about keeping
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...taking care to keep the desired number of replicas...

File: docs/source/install/ewc_ha.rst [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+desired number of replicas to sustain every service alive during the rolling upgrade.
+
+
+Tips & Tricks
+_____________
+Save custom Helm values you want to override in a separated file, upgrade the cluster:
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...in a separate file, ...

File: docs/source/install/ewc_ha.rst [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+.. code-block:: bash
+
+  kubectl logs -l release=<release-name>,tier=backend
+
+
+Custom st2 packs
+----------------
+To follow the stateless model, shipping custom st2 packs is now part of the deployment process.
+It means that ``st2 pack install`` won't work in a distributed environment and you have to bundle all the
+required packs into a Docker image that you could codify, version, package and distribute in a repeatable way.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...that you codify, ...

or

...that you can codify, ...

File: docs/source/install/ewc_ha.rst [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+The responsibility of such Docker image is to hold pack content and their virtualenvs.
+So custom st2 pack Docker image you have to build is just read-only dirs that are shared with the corresponding
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: The custom st2 pack docker image you have to build is essentially a couple read-only directories that are shared with the corresponding st2 services in the cluster.

File: docs/source/install/ewc_ha.rst [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+st2 services in a cluster.
+
+For the convenience we created new ``st2-pack-install <pack1> <pack2> <pack3>`` command
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: For your convenience, we created...

File: docs/source/install/ewc_ha.rst [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+that'll help to install custom packs during the Docker build process without relying on DB and MQ connection.
+
+Helm chart brings helpers to simplify this experience like `stackstorm/st2pack:builder <https://hub.docker.com/r/stackstorm/st2packs/>`_
+Docker image and private Docker registry you can optionally enable in Helm values.yaml to push/pull
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...to easily push/pull your custom packs within the cluster.

File: docs/source/install/ewc_ha.rst [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+  kubectl exec -it ${ST2CLIENT} -- st2 --version
+
+  # switch into a container shell and use st2 CLI
+  kubectl exec -it ${ST2CLIENT} /bin/bash
+
+
+st2web
+______
+st2web is a StackStorm Web UI admin dashboard. By default, st2web K8s config includes a Pod Deployment and a Service.
+``2`` replicas (configurable) of st2web serve the web app and proxify requests to st2auth, st2api, st2stream.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: ...proxy requests...

File: docs/source/install/ewc_ha.rst [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+
+.. note::
+  As Mistral is going to be deprecated and removed from StackStorm platform soon, Helm chart relies only on
+  :doc:`Orquesta st2workflowengine </orquesta/index>` as a new native workflow engine.
+
+st2notifier
+___________
+Multiple st2notifier processes can run in active-active mode, using connections to RabbitMQ and MongoDB and generating triggers based on
+action execution completion as well as doing action rescheduling.
+In an HA deployment minimum ``2`` replicas of st2notifier is running, requiring coordination backend, which is ``etcd`` in our case.
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: In an HA deployment there must be a minimum of ``2`` replicas of st2notifier running, requiring a coordination backend, which in our case is ``etcd``.

File: docs/source/install/ewc_ha.rst [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+
+st2rulesengine
+______________
+st2rulesengine evaluates rules when it sees new triggers and decides if new action execution should be requested.
+K8s config includes Pod Deployment with ``2`` (configurable) replicas by default for HA.
+
+st2timersengine
+_______________
+st2timersengine is responsible for scheduling all user specified `timers <https://docs.stackstorm.com/rules.html#timers>`_ aka st2 cron.
+Only single replica is created via K8s Deployment as timersengine can't work in active-active mode at the moment
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: Only a single replica is created...

File: docs/source/install/ewc_ha.rst (hunk 12 above)
Comment (armab) [reply docs/source/install/ewc_ha.rst]: https://open.nytimes.com/agrarian-scale-kubernetes-part-1-7ea703d4de08

> Visit the Kubernetes homepage, and one of the first things you’ll see is the promise of “Planet Scale.”

How about "of planet scale"?

**Comment by armab** [comment]
@warrenvw Thanks for more review comments, all good stuff!

File: docs/source/install/ewc_ha.rst [hunk 25]
```diff
@@ -0,0 +1,287 @@
...
+
+This document provides an installation blueprint for a Highly Availabile StackStorm Enterprise (|bwc|) cluster
+based on `Kubernetes <https://kubernetes.io/>`__, a container orchestration platform at planet scale.
+
+The cluster deploys minimum of 2 replicas for each component of StackStorm microservices for redundancy and reliability,
+as well as configures backends like MongoDB HA Replicaset, RabbitMQ HA and etcd cluster that st2 relies on for database,
+communication bus, and distributed coordination respectively. That raises a fleet of more than ``30`` pods total.
+
+The source code for K8s resource templates is available as a GitHub repo (TODO):
+`StackStorm/stackstorm-enterprise-ha <https://github.com/StackStorm/stackstorm-enterprise-ha>`_.
```
Comment (armab) [inline docs/source/install/ewc_ha.rst]: ### TODO

- [ ] Merge https://github.com/StackStorm/k8s-st2/pull/78
- [ ] Extract Helm chart to https://github.com/StackStorm/stackstorm-enterprise-ha
- [ ]  Open https://github.com/StackStorm/stackstorm-enterprise-ha repo
- [ ] Copy issues from `k8s-st2` to `stackstorm-enterprise-ha` repo

File: docs/source/install/ewc_ha.rst [hunk 26]
```diff
@@ -124,17 +124,17 @@ Custom st2 packs
...
-required packs into a Docker image that you could codify, version, package and distribute in a repeatable way.
+required packs into a Docker image that you can codify, version, package and distribute in a repeatable way.
 The responsibility of such Docker image is to hold pack content and their virtualenvs.
-So custom st2 pack Docker image you have to build is just read-only dirs that are shared with the corresponding
-st2 services in a cluster.
+So custom st2 pack docker image you have to build is essentially a couple read-only directories that
+are shared with the corresponding st2 services in the cluster.
 
-For the convenience we created new ``st2-pack-install <pack1> <pack2> <pack3>`` command
+For your convenience, we created new ``st2-pack-install <pack1> <pack2> <pack3>`` command
```
Comment (warrenvw) [inline docs/source/install/ewc_ha.rst]: I missed this one before. Should read:  ...we created a new...

File: docs/source/install/ewc_ha.rst (hunk 12 above)
Comment (LindsayHill) [reply docs/source/install/ewc_ha.rst]: I would just not bother with the whole last part. Just link to Kubernetes, let them do their own description.

File: docs/source/install/ewc_ha.rst [hunk 27]
```diff
@@ -0,0 +1,287 @@
...
+
+|bwc| (EWC) is the commercial version of the StackStorm automation platform. EWC adds priority
+support, advanced features such as fine-tuned access control, LDAP, and Workflow Designer. To
+learn more about |bwc|, get an evaluation license, or request a quote, visit `extremenetworks.com/product/workflow-composer
+<https://www.extremenetworks.com/product/workflow-composer/>`_.
+
+This document provides an installation blueprint for a Highly Availabile StackStorm Enterprise (|bwc|) cluster
+based on `Kubernetes <https://kubernetes.io/>`__, a container orchestration platform at planet scale.
+
+The cluster deploys minimum of 2 replicas for each component of StackStorm microservices for redundancy and reliability,
```
Comment (LindsayHill) [inline docs/source/install/ewc_ha.rst]: should be "...deploys a minimum of 2 replicas..."

File: docs/source/install/ewc_ha.rst [hunk 28]
```diff
@@ -0,0 +1,287 @@
+|bwc| HA Cluster - BETA
+=======================
+
+|bwc| (EWC) is the commercial version of the StackStorm automation platform. EWC adds priority
+support, advanced features such as fine-tuned access control, LDAP, and Workflow Designer. To
+learn more about |bwc|, get an evaluation license, or request a quote, visit `extremenetworks.com/product/workflow-composer
+<https://www.extremenetworks.com/product/workflow-composer/>`_.
```
Comment (LindsayHill) [inline docs/source/install/ewc_ha.rst]: Use stackstorm.com/#product here, rather than extremenetworks.com.

**Pull request merged**